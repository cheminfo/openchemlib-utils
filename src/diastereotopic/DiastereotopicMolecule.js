import { getXAtomicNumber } from '../util/getXAtomicNumber.js';
import { makeRacemic } from '../util/makeRacemic.js';
import { tagAtom } from '../util/tagAtom.js';

import { ensureHeterotopicChiralBonds } from './ensureHeterotopicChiralBonds.js';

const MAX_NB_ATOMS = 250;

export class DiastereotopicMolecule {
  constructor(molecule) {
    this.originalMolecule = molecule;
    this.idCode = molecule.getIDCode();
    this.molecule = this.originalMolecule.getCompactCopy();
    this.OCL = molecule.getOCL();
    this.Molecule = this.OCL.Molecule;
    this.molecule.ensureHelperArrays(this.OCL.Molecule.cHelperNeighbours);
    this.cache = {};
  }

  toMolfile() {
    return this.molecule.toMolfile();
  }

  toMolfileWithFinalRanks() {
    return this.moleculeWithFinalRanks.toMolfile();
  }

  finalRankToIndex(finalRank) {
    const finalRanks = this.finalRanks;
    return finalRanks.findIndex((rank) => rank === finalRank);
  }


  /**
   *
   * @param {*} molfile
   */
  updateMolfileWithFinalRanks(molfile) {
    const molecule = this.Molecule.fromMolfile(molfile);
    const idCode = molecule.getIDCode();
    if (idCode !== this.idCode) {
      throw new Error(
        'The molfile does not correspond to the current molecule',
      );
    }
    const xMolecule = getXMolecule(getMoleculeWithH(molecule));
    const finalRanks = getFinalRanks(xMolecule);
    return getMoleculeWithFinalRanks(molecule, finalRanks).toMolfile();
  }

  get moleculeWithH() {
    if (this.cache.moleculeWithH) return this.cache.moleculeWithH;
    this.cache.moleculeWithH = getMoleculeWithH(this.molecule);
    return this.cache.moleculeWithH;
  }

  get xMolecule() {
    if (this.cache.xMolecule) return this.cache.xMolecule;
    this.cache.xMolecule = getXMolecule(this.moleculeWithH);
    return this.cache.xMolecule;
  }

  /**
   * This is related to the current moleculeWithH. The order is NOT canonized
   */
  get diaIDs() {
    if (this.cache.diaIDs) return this.cache.diaIDs;
    this.cache.diaIDs = getDiaIDs(this);
    return this.cache.diaIDs;
  }

  get diaIDsAndH() {
    if (this.cache.diaIDsAndH) return this.cache.diaIDsAndH;
    this.cache.diaIDsAndH = getDiaIDsAndH(this);
    return this.cache.diaIDsAndH;
  }

  /**
   * Returns symmetryRanks for all the atoms including hydrogens. Those ranks
   * deals with topicity and is related to the current moleculeWithH.
   * In order to calculate the ranks we replace all the
   * hydrogens with a X atom.
   */
  get heterotopicSymmetryRanks() {
    if (this.cache.heterotopicSymmetryRanks) return this.cache.heterotopicSymmetryRanks;
    this.cache.heterotopicSymmetryRanks = getHeterotopicSymmetryRanks(this.xMolecule);
    return [...this.cache.heterotopicSymmetryRanks];
  }

  /**
   * Returns finalRanks for all the atoms including hydrogens. Those ranks
   * deals with topicity and is related to the current moleculeWithH.
   * All the atoms have a unique identifier.j
   * In order to calculate the ranks we replace all the
   * hydrogens with a X atom.
   */
  get finalRanks() {
    if (this.cache.finalRanks) return this.cache.finalRanks;
    this.cache.finalRanks = getFinalRanks(this.xMolecule)
    return this.cache.finalRanks;
  }

  get moleculeWithFinalRanks() {
    if (this.cache.moleculeWithFinalRanks) {
      return this.cache.moleculeWithFinalRanks;
    }
    this.cache.moleculeWithFinalRanks = getMoleculeWithFinalRanks(this.molecule, this.finalRanks);
    return this.cache.moleculeWithFinalRanks;
  }

  toMolfileWithH() {
    return this.moleculeWithH.toMolfile();
  }
}

function getMoleculeWithFinalRanks(molecule, finalRanks) {
  const moleculeWithFinalRanks = molecule.getCompactCopy();
  // force hydrogens to be at the end of the file
  moleculeWithFinalRanks.ensureHelperArrays(molecule.getOCL().Molecule.cHelperNeighbours);
  for (let i = 0; i < moleculeWithFinalRanks.getAllAtoms(); i++) {
    moleculeWithFinalRanks.setAtomMapNo(i, finalRanks[i], false);
  }
  return moleculeWithFinalRanks
}


function getDiaIDsAndH(diaMol) {
  const diaIDs = diaMol.diaIDs;
  const newDiaIDs = [];
  const molecule = diaMol.moleculeWithH;

  for (let i = 0; i < diaIDs.length; i++) {
    const diaID = diaIDs[i];
    const newDiaID = {
      oclID: diaID,
      hydrogens: [],
      heavyAtom: undefined,
    };
    if (molecule.getAtomicNo(i) === 1) {
      const atom = molecule.getConnAtom(i, 0);
      newDiaID.heavyAtom = diaIDs[atom];
    }
    for (let j = 0; j < molecule.getAllConnAtoms(i); j++) {
      const atom = molecule.getConnAtom(i, j);
      if (molecule.getAtomicNo(atom) === 1) {
        if (!newDiaID.hydrogens.includes(diaIDs[atom])) {
          newDiaID.hydrogens.push(diaIDs[atom]);
        }
      }
    }
    newDiaIDs.push(newDiaID);
  }
  return newDiaIDs;
}

function getMoleculeWithH(molecule) {
  const moleculeWithH = molecule.getCompactCopy();
  moleculeWithH.addImplicitHydrogens();
  if (moleculeWithH.getAllAtoms() > MAX_NB_ATOMS) {
    throw new Error(
      `Too many atoms to add hydrogens: ${moleculeWithH.getAllAtoms()} > ${MAX_NB_ATOMS}`,
    );
  }
  ensureHeterotopicChiralBonds(moleculeWithH);
  return moleculeWithH;
}



function getDiaIDs(diaMol) {
  const heterotopicSymmetryRanks = diaMol.heterotopicSymmetryRanks;
  const moleculeWithH = diaMol.moleculeWithH;
  const diaIDs = [];
  moleculeWithH.ensureHelperArrays(
    diaMol.Molecule.cHelperSymmetryStereoHeterotopicity,
  );
  const cache = {};
  for (let i = 0; i < diaMol.moleculeWithH.getAllAtoms(); i++) {
    const rank = heterotopicSymmetryRanks[i];
    if (rank && cache[rank]) {
      diaIDs.push(cache[rank].diaID);
      continue;
    }
    const tempMolecule = diaMol.moleculeWithH.getCompactCopy();
    tagAtom(tempMolecule, i);
    makeRacemic(tempMolecule);
    let diaID = tempMolecule.getCanonizedIDCode(
      diaMol.OCL.Molecule.CANONIZER_ENCODE_ATOM_CUSTOM_LABELS,
    );
    diaIDs.push(diaID);
  }
  return diaIDs;
}

function getXMolecule(moleculeWithH) {
  const xAtomNumber = getXAtomicNumber(moleculeWithH);
  const xMolecule = moleculeWithH.getCompactCopy();
  for (let i = 0; i < xMolecule.getAllAtoms(); i++) {
    // hydrogens are not taken into account during canonization, we need to change them with an atom with a valence of 1
    if (xMolecule.getAtomicNo(i) === 1) {
      xMolecule.setAtomicNo(i, xAtomNumber);
    }
  }
  return xMolecule
}

function getHeterotopicSymmetryRanks(xMolecule) {
  xMolecule.ensureHelperArrays(
    xMolecule.getOCL().Molecule.cHelperSymmetryStereoHeterotopicity,
  );
  const symmetryRanks = [];
  for (let i = 0; i < xMolecule.getAllAtoms(); i++) {
    symmetryRanks.push(xMolecule.getSymmetryRank(i));
  }
  return symmetryRanks;
}

function getFinalRanks(xMolecule) {
  xMolecule.ensureHelperArrays(
    xMolecule.getOCL().Molecule.cHelperSymmetryStereoHeterotopicity,
  );
  return [...xMolecule.getFinalRanks()];
}


