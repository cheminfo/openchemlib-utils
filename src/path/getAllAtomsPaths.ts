import { Molecule } from 'openchemlib';

export interface GetAllAtomsPathsOptions {
  /**
   * Maximum path length to consider. All the paths from 0 to maxPathLength will be considered
   * @default 5
   */
  maxPathLength?: number;
}

interface AtomPath {
  path: number[];
  distance: number;
}

/**
 * We need to create an array of atoms
 * that contains an array of pathLength
 * that contains an array of object
 * @param molecule
 * @param options
 * @returns
 */
export function getAllAtomsPaths(
  molecule: Molecule,
  options: GetAllAtomsPathsOptions = {},
) {
  const { maxPathLength = 5 } = options;
  const allAtomsPaths = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    const oneAtomPaths: AtomPath[][] = [];
    allAtomsPaths.push(oneAtomPaths);

    let atomPaths: AtomPath[] = [];
    atomPaths.push({ path: [i], distance: 0 });
    oneAtomPaths.push(atomPaths);

    let nextIndexes = [0];
    let nextAtoms = [i];

    for (let sphere = 1; sphere <= maxPathLength; sphere++) {
      atomPaths = [];
      oneAtomPaths.push(atomPaths);
      const currentIndexes = nextIndexes;
      const currentAtoms = nextAtoms;
      nextIndexes = [];
      nextAtoms = [];

      for (let i = 0; i < currentIndexes.length; i++) {
        const atom = currentAtoms[i];
        const index = currentIndexes[i];

        const previousPath = oneAtomPaths[sphere - 1][index].path;
        for (let conn = 0; conn < molecule.getConnAtoms(atom); conn++) {
          const connectedAtom = molecule.getConnAtom(atom, conn);
          if (previousPath.includes(connectedAtom)) continue;
          nextIndexes.push(atomPaths.length);
          nextAtoms.push(connectedAtom);
          atomPaths.push({
            path: [...previousPath, connectedAtom],
            distance: sphere,
          });
        }
      }
    }
  }
  return allAtomsPaths;
}
