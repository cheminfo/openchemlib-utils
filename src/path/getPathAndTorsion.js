/**
 * Calculates the path between 2 atoms
 * @param {import('openchemlib').Molecule} molecule
 * @param {number} from - index of the first atom
 * @param {number} to - index of the end atom
 * @param {number} maxLength - maximal length of the path
 */
export function getPathAndTorsion(molecule, from, to, maxLength) {
  let originalAtoms = []; // path before renumbering
  molecule.getPath(originalAtoms, from, to, maxLength + 1);
  let torsion;
  if (originalAtoms.length === 4) {
    torsion = molecule.calculateTorsion(originalAtoms);
  }

  return {
    atoms: originalAtoms,
    from,
    to,
    torsion,
    length: originalAtoms.length - 1,
  };
}
