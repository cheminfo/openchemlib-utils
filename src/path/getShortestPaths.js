/**
 * Get the shortest path between each pair of atoms in the molecule
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.maxLength=4]
 * @returns {Array<Array>} A matrix containing on each cell (i,j) the shortest path from atom i to atom j
 */
export function getShortestPaths(molecule, options = {}) {
  const OCL = molecule.getOCL();
  const { fromLabel = '', toLabel = '', maxLength = 3 } = options;

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  const nbAtoms = molecule.getAllAtoms();
  let allShortestPaths = new Array(nbAtoms);
  for (let i = 0; i < nbAtoms; i++) {
    allShortestPaths[i] = new Array(nbAtoms);
  }

  for (let from = 0; from < nbAtoms; from++) {
    allShortestPaths[from][from] = [from];
    for (let to = from + 1; to < nbAtoms; to++) {
      if (
        (fromAtomicNumber === 0 ||
          molecule.getAtomicNo(from) === fromAtomicNumber) &&
        (toAtomicNumber === 0 || molecule.getAtomicNo(to) === toAtomicNumber)
      ) {
        let path = [];
        molecule.getPath(path, from, to, maxLength);
        if (path.length) {
          allShortestPaths[from][to] = path.slice();
          allShortestPaths[to][from] = path.reverse();
        } else {
          allShortestPaths[from][to] = null;
          allShortestPaths[to][from] = null;
        }
      } else {
        allShortestPaths[from][to] = null;
        allShortestPaths[to][from] = null;
      }
    }
  }

  return allShortestPaths;
}
