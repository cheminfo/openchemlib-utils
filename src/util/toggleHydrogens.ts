import type { Molecule } from 'openchemlib';

import { getImplicitHydrogensCount } from './getImplicitHydrogensCount';

/**
 * Toggle presence of implicity hydrogens on/off
 * @param molecule
 * @param atomID
 */
export function toggleHydrogens(molecule: Molecule, atomID: number) {
  if (getImplicitHydrogensCount(molecule, atomID) === 0) {
    const atomsToDelete = [];
    for (let i = 0; i < molecule.getAllConnAtoms(atomID); i++) {
      const connectedAtom = molecule.getConnAtom(atomID, i);
      if (molecule.getAtomicNo(connectedAtom) === 1) {
        atomsToDelete.push(connectedAtom);
      }
    }
    molecule.deleteAtoms(atomsToDelete);
  } else {
    molecule.addImplicitHydrogens(atomID);
  }
}
