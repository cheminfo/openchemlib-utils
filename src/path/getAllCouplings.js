import { tagAtom } from '../util/tagAtom';
import { makeRacemic } from '../util/makeRacemic';

/**
 * Returns an array of all the different atom diaIDs that are connected
 * {OCL}
 * {OCL.Molecule} molecule
 * {object} [options={}]
 * {string} [options.fromLabel='H']
 * {string} [options.toLabel='H']
 * {number} [options.minLength=1]
 * {number} [options.maxLength=4]
 */

function getAllCouplings(OCL, molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;
  let paths = molecule.getAllPaths({
    fromLabel,
    toLabel,
    minLength,
    maxLength,
  });

  const minSphereSize = 0;
  const maxSphereSize = 2;

  let fragment = new OCL.Molecule(0, 0);
  for (let path of paths) {
    path.info = [];
    for (let fromTo of path.fromTo) {
      let atoms = [];
      molecule.getPath(atoms, fromTo[0], fromTo[1], path.pathLength);
      let torsion;
      if (atoms.length === 4) {
        torsion = molecule.calculateTorsion(atoms);
      }
      path.info.push({
        atoms,
        torsion,
      });

      if (!path.code) {
        path.code = [];

        let tmpMolecule = molecule.getCompactCopy();
        makeRacemic(OCL, tmpMolecule);
        tagAtom(OCL, tmpMolecule, atoms[0]);
        tagAtom(OCL, tmpMolecule, atoms[atoms.length - 1]);

        let atomMask = new Array(tmpMolecule.getAllAtoms()).fill(false);
        let atomList = [];
        let max = 0;
        let min = 0;
        for (let sphere = 0; sphere <= maxSphereSize; sphere++) {
          if (max === 0) {
            for (let atom of atoms) {
              atomMask[atom] = true;
              atomList.push(atom);
              max++;
            }
          } else {
            let newMax = max;
            for (let i = min; i < max; i++) {
              let atom = atomList[i];
              for (let j = 0; j < tmpMolecule.getConnAtoms(atom); j++) {
                let connAtom = tmpMolecule.getConnAtom(atom, j);
                atomMask[connAtom] = true;
                atomList[newMax++] = connAtom;
              }
            }
            min = max;
            max = newMax;
          }
          tmpMolecule.copyMoleculeByAtoms(fragment, atomMask, true, null);
          if (sphere >= minSphereSize) {
            path.code.push(
              fragment.getCanonizedIDCode(
                OCL.Molecule.CANONIZER_ENCODE_ATOM_CUSTOM_LABELS,
              ),
            );
          }
        }
      }
    }
  }
  return paths;
}

module.exports = getAllCouplings;
