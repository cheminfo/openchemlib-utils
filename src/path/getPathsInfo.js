import { getHoseCodesForPath } from '../hose/getHoseCodesForPath';
import { getAtomsInfo } from '../util/getAtomsInfo';
import { getConnectivityMatrix } from '../util/getConnectivityMatrix';

let fragment;

/**
 *
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.minLength=1]
 * @param {string} [opions.maxLength=4]

 */
export function getPathsInfo(molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;

  const OCL = molecule.getOCL();

  if (!fragment) {
    fragment = new OCL.Molecule(0, 0);
  }

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  // we need to find all the atoms 'fromLabel' and 'toLabel'
  let atomsInfo = getAtomsInfo(molecule);

  let pathLengthMatrix = getConnectivityMatrix(molecule, {
    pathLength: true,
  });

  for (let from = 0; from < molecule.getAllAtoms(); from++) {
    atomsInfo[from].paths = [];
    for (let to = 0; to < molecule.getAllAtoms(); to++) {
      if (from !== to) {
        if (molecule.getAtomicNo(from) === fromAtomicNumber) {
          if (molecule.getAtomicNo(to) === toAtomicNumber) {
            let pathLength = pathLengthMatrix[from][to];
            if (pathLength >= minLength && pathLength <= maxLength) {
              atomsInfo[from].paths.push(
                getHoseCodesForPath(
                  molecule,
                  from,
                  to,
                  pathLength,
                  atomsInfo[to].oclID,
                ),
              );
            }
          }
        }
      }
    }
  }

  return atomsInfo;
}
