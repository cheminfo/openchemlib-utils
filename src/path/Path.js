import { getPathsInfo } from './getPathsInfo';

export const Path = {
  getPathsInfo,
};
