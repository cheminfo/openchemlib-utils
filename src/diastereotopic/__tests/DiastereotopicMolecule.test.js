import { readFileSync } from 'fs';
import { join } from 'path';

import { Molecule } from 'openchemlib';

import { toggleHydrogens } from '../../util/toggleHydrogens.ts';
import { DiastereotopicMolecule } from '../DiastereotopicMolecule.js';

describe('DiastereotopicMolecule', () => {
  it('ethanol', () => {
    const molecule = Molecule.fromSmiles('CCCO');
    molecule.setAtomicNo(0, 1);
    const diastereotopicMolecule = new DiastereotopicMolecule(molecule);
    const diaIDs = diastereotopicMolecule.diaIDs;
    expect(diaIDs).toHaveLength(9);
    expect(diaIDs).toStrictEqual([
      'eMHAIhFHhOtdGrj@',
      'eMHAIhFIhOtdWBj@',
      'eMHAIhFJhOtdgBj@',
      'gCaHLIeIZ`GzQ@bUP',
      'gCaHHIeIZ`GzQ@bUP',
      'gCaHLIeIZ`GzQ@bUP',
      'gCaHLIeIZ`GzQ@bUP',
      'gCaHDIeIjiJ@\x7FRHDRj@',
      'gCaHDIeIjiJ@\x7FRHDRj@',
    ]);
    const diaIDsAndH = diastereotopicMolecule.diaIDsAndH;
    expect(diaIDsAndH).toHaveLength(9);
    expect(diaIDsAndH).toMatchSnapshot()
    const molfile = diastereotopicMolecule.toMolfile();
    expect(getMolfileAtoms(molfile)).toStrictEqual(['O', 'C', 'C', 'H']);
    const molfileWithH = diastereotopicMolecule.toMolfileWithH();
    expect(getMolfileAtoms(molfileWithH)).toStrictEqual([
      'O',
      'C',
      'C',
      'H',
      'H',
      'H',
      'H',
      'H',
      'H',
    ]);
    const molfileWithAtomMapNo =
      diastereotopicMolecule.toMolfileWithFinalRanks();
    expect(getMolfileAtoms(molfileWithAtomMapNo)).toStrictEqual([
      'O',
      'C',
      'C',
      'H',
    ]);
    expect(getMolfileAtomMapNo(molfileWithAtomMapNo)).toStrictEqual([
      '3',
      '2',
      '1',
      '8',
    ]);

    function getNewMolecule() {
      const molecule = Molecule.fromSmiles('CCCC')
      molecule.setAtomicNo(2, 8);
      molecule.setAtomicNo(3, 1);
      const newMolfile = diastereotopicMolecule.updateMolfileWithFinalRanks(molecule.toMolfile());
      return Molecule.fromMolfile(newMolfile);
    }

    const newMolecule = getNewMolecule();
    const atoms = []
    for (let i = 0; i < newMolecule.getAllAtoms(); i++) {
      const atom = {
        atomicNo: newMolecule.getAtomicNo(i),
        ...diastereotopicMolecule.diaIDsAndH[diastereotopicMolecule.finalRankToIndex(newMolecule.getAtomMapNo(i))]
      }
      atoms.push(atom);
    }

    expect(atoms).toHaveLength(4);
    expect(atoms).toMatchSnapshot();
  });

  it('ethanol toggle implicit H', () => {
    const molecule = Molecule.fromSmiles('CCO');
    const diastereotopicMolecule = new DiastereotopicMolecule(molecule)

    toggleHydrogens(molecule, 0)
    let newMolecule = Molecule.fromMolfile(diastereotopicMolecule.updateMolfileWithFinalRanks(molecule.toMolfile()));
    let atoms = []
    for (let i = 0; i < newMolecule.getAllAtoms(); i++) {
      const atom = {
        atomicNo: newMolecule.getAtomicNo(i),
        ...diastereotopicMolecule.diaIDsAndH[diastereotopicMolecule.finalRankToIndex(newMolecule.getAtomMapNo(i))]
      }
      atoms.push(atom);
    }
    toggleHydrogens(molecule, 1)
    toggleHydrogens(molecule, 0)
    newMolecule = Molecule.fromMolfile(diastereotopicMolecule.updateMolfileWithFinalRanks(molecule.toMolfile()));
    atoms = []
    for (let i = 0; i < newMolecule.getAllAtoms(); i++) {
      const atom = {
        atomicNo: newMolecule.getAtomicNo(i),
        ...diastereotopicMolecule.diaIDsAndH[diastereotopicMolecule.finalRankToIndex(newMolecule.getAtomMapNo(i))]
      }
      atoms.push(atom);
    }
  })


  it.skip('2-chlorobutane', () => {
    const molecule = Molecule.fromSmiles('C[C@H](Cl)CC');
    const diastereotopicMolecule = new DiastereotopicMolecule(molecule);

    console.log(diastereotopicMolecule.heterotopicSymmetryRanks)
    console.log(diastereotopicMolecule.finalRanks)



  });

  it.skip('cyclosporin', () => {
    const molfile = readFileSync(
      join(__dirname, 'data/cyclosporin.mol'),
      'utf8',
    );
  });
});

function getMolfileAtoms(molfile) {
  const lines = molfile.split(/\r?\n/);
  const atoms = [];
  for (let line of lines) {
    if (line.match(/ [A-Z][a-z]? /)) {
      atoms.push(line.replace(/^.* ([A-Z][a-z]?) .*$/, '$1'));
    }
  }
  return atoms;
}

function getMolfileAtomMapNo(molfile) {
  const lines = molfile.split(/\r?\n/);
  const atomMapNos = [];
  for (let line of lines) {
    if (line.match(/ [A-Z][a-z]? /)) {
      atomMapNos.push(line.replace(/^.* ([0-9]+) {2}0 {2}0$/, '$1'));
    }
  }
  return atomMapNos;
}
