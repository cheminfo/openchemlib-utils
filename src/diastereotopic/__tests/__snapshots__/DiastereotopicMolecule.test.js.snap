// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DiastereotopicMolecule > ethanol 1`] = `
[
  {
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHHIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFHhOtdGrj@",
  },
  {
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHLIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFIhOtdWBj@",
  },
  {
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHDIeIjiJ@RHDRj@",
    ],
    "oclID": "eMHAIhFJhOtdgBj@",
  },
  {
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFHhOtdGrj@",
    "hydrogens": [],
    "oclID": "gCaHHIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFIhOtdWBj@",
    "hydrogens": [],
    "oclID": "gCaHLIeIZ\`GzQ@bUP",
  },
  {
    "heavyAtom": "eMHAIhFJhOtdgBj@",
    "hydrogens": [],
    "oclID": "gCaHDIeIjiJ@RHDRj@",
  },
  {
    "heavyAtom": "eMHAIhFJhOtdgBj@",
    "hydrogens": [],
    "oclID": "gCaHDIeIjiJ@RHDRj@",
  },
]
`;

exports[`DiastereotopicMolecule > ethanol 2`] = `
[
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHLIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFIhOtdWBj@",
  },
  {
    "atomicNo": 6,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHDIeIjiJ@RHDRj@",
    ],
    "oclID": "eMHAIhFJhOtdgBj@",
  },
  {
    "atomicNo": 8,
    "heavyAtom": undefined,
    "hydrogens": [
      "gCaHHIeIZ\`GzQ@bUP",
    ],
    "oclID": "eMHAIhFHhOtdGrj@",
  },
  {
    "atomicNo": 1,
    "heavyAtom": "eMHAIhFHhOtdGrj@",
    "hydrogens": [],
    "oclID": "gCaHHIeIZ\`GzQ@bUP",
  },
]
`;
